torch/distributed/nn/_functions.py

```diff
@@ -0,0 +1,339 @@
+        # aggregation, get it from the group size
+        # tensor should be correctly sized for the method
+        # gathering
+        tensor_list = [
+            torch.empty_like(tensor) for i in range(dist.get_world_size())
+        ]
+        ctx.rank = dist.get_rank()
+        dist.gather(tensor, tensor_list, src, group=group)
+        return tuple(tensor_list)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None) + (_Scatter.apply(ctx.src, ctx.group, *grad_outputs),)
+
+
+class _Scatter(Function):
+    @staticmethod
+    def forward(ctx, src, group, *tensors):
+        ctx.src = src
+        ctx.group = group
+        # TODO assert all the tensors have the same size
+        output = torch.empty_like(tensors[0])
+        dist.scatter(output, tensors, src, group=group)
+        return output
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None) + _Gather.apply(ctx.src, ctx.group, *grad_outputs)
+
+
+class _Reduce(Function):
+    @staticmethod
+    def forward(ctx, src, op, group, tensor):
+        ctx.src = src
+        ctx.group = group
+        dist.reduce(tensor, src, op=op, group=group)
+        return tensor
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None, None) + (_Broadcast.apply(ctx.src, ctx.group, *grad_outputs),)
+
+
+class _ReduceMultiGPU(Function):
+    @staticmethod
+    def forward(ctx, dst, op, group, dst_tensor, *tensors):
+        ctx.dst = dst
+        ctx.group = group
+        dist.reduce_multigpu(tensors, dst, op=op, group=group)
+        return tuple(tensors)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None, None, None) + _BroadcastMultiGPU.apply(ctx.src, ctx.group, *grad_outputs)
+
+
+
+class _AllGather(Function):
+    @staticmethod
+    def forward(ctx, group, tensor):
+        ctx.group = group
+        out_tensor_list = [
+            torch.empty_like(tensor) for i in range(dist.get_world_size())
+        ]
+        dist.all_gather(out_tensor_list, tensor, group=group)
+        return tuple(out_tensor_list)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        gxs = _AlltoAll.apply(ctx.group, *grad_outputs)
+        gx = torch.stack(gxs).sum(dim=0)
+        return (None, gx)
+
+
+class _AlltoAll(Function):
+    @staticmethod
+    def forward(ctx, group, *tensors):
+        ctx.group = group
+        out_tensor_list = [
+            torch.empty_like(tensor) for i in range(dist.get_world_size())
+        ]
+        dist.all_to_all(out_tensor_list, tensors, group=group)
+        return tuple(out_tensor_list)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None) + _AlltoAll.apply(ctx.group, *grad_outputs)
+
+
+class _AllReduce(Function):
+    @staticmethod
+    def forward(ctx, op, group, tensor):
+        ctx.group = group
+        ctx.op = op
+        tensor = tensor.clone()
```

Comment:
What is the contract for this function? Should it modify the input inplace?
- If so, you can do it as long as you mark it with `ctx.mark_dirty(tensor)`.
- Users might get in trouble if the original value was needed and they will get the "a variable needed for backward was modified inplace error" so it is a tradeoff
- If you do so, you need to be extra careful in the backward formulas as you are *never* allowed to modified the given grad_outputs.

Reply:
Didn't know this, great catch!